client/src/components/PopoverOptionSet/PopoverOptionSet.js

```diff
@@ -0,0 +1,238 @@
+ */
+class PopoverOptionSet extends Component {
+  constructor(props) {
+    super(props);
+
+    this.handleToggle = this.handleToggle.bind(this);
+    this.handleSearchValueClear = this.handleSearchValueClear.bind(this);
+    this.handleSearchValueChange = this.handleSearchValueChange.bind(this);
+    this.handleKeyDown = this.handleKeyDown.bind(this);
+
+    this.state = {
+      searchValue: ''
+    };
+  }
+
+  /**
+   * Pass toggle to parent and clear the search input
+   */
+  handleToggle() {
+    const { toggle } = this.props;
+
+    toggle();
+    this.handleSearchValueClear();
+  }
+
+  /**
+   * Handle click on clear button within search bar
+   */
+  handleSearchValueClear() {
+    this.setState(
+      { searchValue: '' }
+    );
+  }
+
+  /**
+   * Update the internal state on user input change
+   * @param {Event} event
+   */
+  handleSearchValueChange(event) {
+    this.setState(
+      { searchValue: event.target.value }
+    );
+  }
+
+  handleKeyDown(event) {
+    if (
+      // Ignore key presses in the search field...
+      event.target
+      && event.target.nodeName === 'INPUT'
+      // ...but allow "escape"
+      && event.key !== 'Escape'
+    ) {
+      return;
+    }
+
+    switch (event.key) {
+      case 'Escape':
+        this.handleToggle();
+        break;
+      default:
+    }
+  }
+
+  /**
+   * Render a link to clear the search field if user entered input
+   * @returns {DOMElement}
+   */
+  renderSearchValueClearLink() {
+    const { searchValue } = this.state;
+
+    if (searchValue.length === 0) {
+      return null;
+    }
+
+    return (
+      <InputGroupAddon addonType="append">
+        <button
+          className="popover-option-set__search-clear btn-link"
+          onClick={this.handleSearchValueClear}
+        >
+          {i18n._t('PopoverOptionSet.CLEAR', 'Clear')}
+        </button>
+      </InputGroupAddon>
+    );
+  }
+
+  /**
+   * Render the search value intput box (which is in turn used to filter the buttons)
+   *
+   * @return {DOMElement}
+   */
+  renderSearchBox() {
+    const { searchPlaceholder, disableSearch } = this.props;
+    const { searchValue } = this.state;
+    if (disableSearch) {
+      return null;
+    }
+    return (
+      <InputGroup className="popover-option-set__search">
+        <Input
+          autoFocus
+          className="popover-option-set__search-input"
+          id="popover-option-set__search-input"
+          onChange={this.handleSearchValueChange}
+          placeholder={searchPlaceholder}
+          type="text"
+          value={searchValue}
+        />
+        {this.renderSearchValueClearLink()}
+      </InputGroup>
+    );
+  }
+
+  /**
+   * Render either all buttons available, buttons matching the search term, or a message that
+   * there are no matching buttons
+   * @returns {DOMElement}
+   */
+  renderOptionButtons() {
+    const { onSearch, onButtonClick, ButtonComponent } = this.props;
+    let { buttons } = this.props;
```

Comment:
We shouldn't be mutating props. Perhaps you could do something like this instead:

```js
const { buttons } = this.props;

let buttonsToRender = buttons;
if (...) {
  buttonsToRender = 'something else';
```

Reply:
Sure. `buttons` is an array and will not be taken by reference so it won't mutate. I can add the extra line but it would only be for clarity I guess.

Reply:
I think it'd be worth it, otherwise someone might see that buttonsToRender is mutable and do it